File: frontend/src/lib/spindoctorRefs.ts
const REF_PATTERN = /\[\[ref:([a-z0-9_-]+)\]\]/gi

/** Strip Spinndoktor section markers from assistant text shown in the thread. */
export function stripSpindoctorRefs(text: string): string {
  return text.replace(REF_PATTERN, "").trim()
}

/** Last [[ref:id]] in a reply, if any. */
export function lastSpindoctorRef(text: string): string | null {
  const matches = [...text.matchAll(new RegExp(REF_PATTERN.source, "gi"))]
  if (matches.length === 0) return null
  return matches[matches.length - 1][1] ?? null
}

export function scrollReportCanvas(
  iframe: HTMLIFrameElement | null,
  sectionId: string,
): void {
  if (!iframe?.contentWindow) return
  iframe.contentWindow.postMessage(
    { type: "spinndoctor-scroll", id: sectionId },
    "*",
  )
}

File: backend/app/services/spindoctor_refs.py
"""Parse Spinndoktor [[ref:id]] markers from assistant replies."""

from __future__ import annotations

import re

REF_PATTERN = re.compile(r"\[\[ref:([a-z0-9_-]+)\]\]", re.IGNORECASE)


def strip_spindoctor_refs(text: str) -> str:
    return REF_PATTERN.sub("", text).strip()


def last_spindoctor_ref(text: str) -> str | None:
    matches = list(REF_PATTERN.finditer(text))
    if not matches:
        return None
    return matches[-1].group(1)
